pymc3/variational/replacements.py

```diff
@@ -0,0 +1,285 @@
+        return collections.OrderedDict(
+            [(self.model[k], v) for k, v in self.local_dict['means'].items()]
+        )
+
+    @property
+    def deterministic_replacements_global(self):
+        raise NotImplementedError
+
+    @property
+    def params(self):
+        """
+        Returns
+        -------
+        list - shared params to fit
+        """
+        return self.params_local + self.params_global
+
+    @property
+    def params_local(self):
+        """
+        Returns
+        -------
+        list - shared params for local replacements
+        """
+        return []
+
+    @property
+    def params_global(self):
+        """
+        Returns
+        -------
+        list - shared params for global replacements
+        """
+        raise NotImplementedError
+
+    @property
+    def log_q_W_local(self):
+        if self.local_dict['x']:
+            x = flatten_list(self.local_dict['x'])
+            mu = flatten_list(self.local_dict['means'].values())
+            rho = flatten_list(self.local_dict['rhos'].values())
+            _log_q_W_local_ = tt.sum(log_normal3(x, mu, rho))
+            return _log_q_W_local_
+        else:
+            return 0
+
+    @property
+    def log_q_W_global(self):
+        raise NotImplementedError
+
+    @property
+    def log_q_W(self):
+        return self.log_q_W_global + self.log_q_W_local
+
+    @property
+    def log_p_D(self):
+        _log_p_D_ = tt.add(
+            *map(self.weighted_likelihood, self.model.observed_RVs)
+        )
+        return _log_p_D_
+
+    @property
+    def log_p_W(self):
+        _log_p_W_ = self.model.varlogpt + tt.sum(self.model.potentials)
+        return _log_p_W_
+
+    @property
+    def elbo(self):
+        return self.log_p_D + self.log_p_W - self.log_q_W
+
+    def apply_replacements(self, node, deterministic=False):
+        if deterministic:
+            return theano.clone(
+                node, self.deterministic_replacements, strict=False)
+        else:
+            return theano.clone(
+                node, self.stochastic_replacements, strict=False)
+
+    def weighted_likelihood(self, var):
+        tot = self.population.get(
+            var, self.population.get(var.name))
+        logpt = tt.sum(var.logpt)
+        if tot is not None:
+            tot = tt.as_tensor(tot)
+            logpt *= tot
+            logpt /= var.size
+        return logpt
+
+    def sample_elbo(self, samples=1, pi=1):
+        """Output of this function should be used for fitting a model
+
+        Parameters
+        ----------
+        samples : Tensor - number of samples
+        pi : Tensor - weight of variational part
+
+
+        Notes
+        -----
+        In case of samples == 1 updates are empty and can be ignored
+
+        Returns
+        -------
+        elbos, updates
+        """
+        samples = tt.as_tensor(samples)
+        pi = tt.as_tensor(pi)
+        _elbo_ = self.log_p_D + pi * (self.log_p_W - self.log_q_W)
```

Comment:
In the reference, `pi` is defined for global RVs. But when using AEVB, both of `log_p_W` and `log_q_W` include global and local RVs. I thinks the local RV terms should not be weighted by `pi`. 

```
 +    @property
 +    def log_q_W_local(self):
 +        if self.local_dict['x']:
 +            x = flatten_list(self.local_dict['x'])
 +            mu = flatten_list(self.local_dict['means'].values())
 +            rho = flatten_list(self.local_dict['rhos'].values())
 +            _log_q_W_local_ = tt.sum(log_normal3(x, mu, rho))
 +            return _log_q_W_local_
 +        else:
 +            return 0
 +
 +    @property
 +    def log_q_W_global(self):
 +        raise NotImplementedError
 +
 +    @property
 +    def log_q_W(self):
 +        return self.log_q_W_global + self.log_q_W_local
```
The above code weights global and local RVs the same amount.